feat: Add custom_fetch to allow alternate WebAPI Fetch implementation

agrohs opened the pull request:
## Type

* ### Feature
  Implements a new feature

## Description
_A brief description of the goals of the pull request._
* Allows `custom_fetch` to be passed into `MoltinGateway` configuration that will be used for authentication calls to API. This allows alternate implementations of the WebAPI Fetch (https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) spec to be used.

```
    const Moltin = MoltinGateway({
      client_id: 'X',
      client_secret: 'X',
      grant_type: 'client_credentials',
      custom_fetch: myCustomWebAPIfetch
    })
```

## Dependencies
_Other PRs or builds that this PR depends on._
* N/A

## Issues

*A list of issues closed by this PR.*

* N/A

## Notes

* Ex. - a scenario where you want to proxy the request for a `client_credentials` token to avoid exposing the `client_secret` in any requests.

agrohs commented:
@gje4 - just an FYI, this is the PR I was mentioning to support the secure WebAPI Fetch implementation

ynnoj commented:
@agrohs Thanks for the pull request. 

I’m struggling to understand a scenario when passing in another fetch interface would be required. Can you describe the use case in your situation, perhaps provide some implementation examples?

Thanks!

agrohs commented:
No worries, I figured it might come out of context. In the example in the Notes at the bottom of the PR, described the scenario that we are looking at (though I could imagine a case where others might have valid reasons to have their own custom implementation). In our case, we are simply sending requests to a proxy that has access to private keys and contains a token replacement system it uses before they are sent to the original requested URL (with all headers and such intact). This ensure the client secret is not exposed on the client side, but are replaced before the proxied fetch from there goes through to the original Moltin API call. Happy to hop on a call to talk through and show you the custom implementation we are sending in/why we need it?

George has a bit more background before this was submitted and all has been tested on our end using the custom Fetch implementation to route through the proxy.


ADAM GROHS.
CO-FOUNDER/CEO
646.244.4501

Particular.
the innovation agency of tomorrow, today

www.uniquelyparticular.com

> On Mar 17, 2019, at 12:25 PM, Jonathan Steele <notifications@github.com> wrote:
> 
> @agrohs <https://github.com/agrohs> Thanks for the pull request.
> 
> I’m struggling to understand a scenario when passing in another fetch interface would be required. Can you describe the use case in your situation, perhaps provide some implementation examples?
> 
> Thanks!
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/moltin/js-sdk/pull/248#issuecomment-473686720>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AAJzhyo9uWwh1caTQaWU9iNI8UnR-fcEks5vXnqmgaJpZM4b4dQt>.
>

ynnoj commented:
> In our case, we are simply sending requests to a proxy that has access to private keys and contains a token replacement system it uses before they are sent to the original requested URL (with all headers and such intact)

If there’s already some server element involved, why not just make the requests that require secret keys from there and query that server from the client?

agrohs commented:
We don’t have any sever element on our end, but a product it’s being wrapped in that provides a proxy method to secure keys for this exact purpose. We’ve simply wrapped their proxy in an implementation of the Fetch spec to make it easier to plug in here (and elsewhere).

www.uniquelyparticular.com

> On Mar 17, 2019, at 1:04 PM, Jonathan Steele <notifications@github.com> wrote:
> 
> In our case, we are simply sending requests to a proxy that has access to private keys and contains a token replacement system it uses before they are sent to the original requested URL (with all headers and such intact)
> 
> If there’s already some server element involved, why not just make the requests that require secret keys from there and query that server from the client?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

ynnoj commented:
@agrohs What does the `myCustomWebAPIfetch` in your example look like? Which fields do you need to override that are implemented by the standard fetch interface in the auth request?

notrab commented on src/factories/request.js:
```diff
@@ -70,6 +71,7 @@ class RequestFactory {
         config.client_id,
         response.access_token,
         response.expires
+        // Math.min(response.expires, config.auth.expires)
```
@agrohs Is this required?

agrohs replied:
This is actually related to what I believe may be a defect in the SDK, I can remove the commented out block to leave as is but love to discuss as I think the current implementation may have an issue.

agrohs replied:
Committed removal of comment about potential issue around credential expiration.

notrab commented on src/config.js:
```diff
@@ -9,7 +9,8 @@ class Config {
       host,
-      storage
+      storage,
+      secure_fetch
```
@agrohs @ynnoj is `secure_fetch` the best name for this param? Would `custom_fetch` work?

agrohs replied:
Yes, I’m not partial as to what it’s called (actually had as something like custom_fetch earlier, but as it is used for the auto call was playing with auth_fetch or secure_fetch or credentials_fetch).

agrohs replied:
Updated name and committed as custom_fetch

notrab commented:
@agrohs Are you able to provide a basic outline of `myCustomWebAPIfetch`? I can use that in writing a test and get this merged.

agrohs commented:
Yep, I’m not partial as to what it’s called (actually had as something like custom_fetch earlier, but as it is used for the auto call was playing with auth_fetch or secure_fetch or credentials_fetch).

ADAM GROHS.
CO-FOUNDER/CEO

Particular.
the innovation agency of tomorrow, today

www.uniquelyparticular.com

> On Mar 27, 2019, at 5:19 AM, Jamie Barton <notifications@github.com> wrote:
> 
> @notrab commented on this pull request.
> 
> In src/config.js:
> 
> > @@ -9,7 +9,8 @@ class Config {
>        client_secret,
>        currency,
>        host,
> -      storage
> +      storage,
> +      secure_fetch
> @agrohs @ynnoj is secure_fetch the best name for this param? Would custom_fetch work?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

agrohs commented:
Thanks @notrab, I'm happy to share parts of our implementation offline if helpful, but may be best from a test case side to just try one of the custom implementations on npm (https://www.npmjs.com/package/node-fetch <https://github.com/github/fetch>) or the standard fetch polyfill (https://github.com/github/fetch <https://github.com/github/fetch>). There may even be test cases in one of those repos that would be helpful to grab for a basic nock or something?

> On Mar 27, 2019, at 5:20 AM, Jamie Barton <notifications@github.com> wrote:
> 
> @agrohs <https://github.com/agrohs> Are you able to provide a basic outline of myCustomWebAPIfetch? I can use that in writing a test and get this merged.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/moltin/js-sdk/pull/248#issuecomment-477081205>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AAJzh2ZHIoVuxEZil-597KuugltS-CR-ks5va0X5gaJpZM4b4dQt>.
>

agrohs commented:
Just checking in on this PR @notrab, @gje4, @ynnoj?

agrohs commented:
@notrab, committed the additional changes/tests we discussed into the fork

agrohs commented:
added missing catch clause on authenticate promise (need to double check, but currently have this deleting `moltinCredentials` from storage - could also just leave storage as is and do nothing in the catch clause, but seems if there is an auth error, that we would want to remove the previous stored credentials.

also, added the missing test that checks to ensure that if an invalid custom_fetch is passed in that it is rejected as a TypeError!

cc: @gje4, @notrab

notrab commented:
Looks good to me!

notrab commented on src/factories/request.js:
```diff
@@ -65,14 +66,18 @@ class RequestFactory {
+        storage.set('moltinCredentials', JSON.stringify(credentials))
+      })
+      .catch(() => {
+        storage.delete('moltinCredentials')
```
@agrohs Could you remove this? We can then apply this in a separate PR.

agrohs replied:
@notrab, shouldn't remove the catch itself as it is throwing an `UnhandledPromiseRejection` when a bad fetch is passed in (as tested in the unit test), but can remove the `storage.delete('moltinCredentials')` inside and you guys can figure out whether that is a desired fix/feature to add in on the moltin end at some point. Will update commit now w/ the empty catch block now per our discussion.
